File: Difficulty: Medium/Count all triplets with given sum in sorted array/count-all-triplets-with-given-sum-in-sorted-array.py

class Solution:
    def countTriplets(self, arr, target):
        # code here
        def count(k, target):
            nonlocal arr
            cnt = 0
            i, j = k, len(arr)-1
            while i < j:
                t = arr[i] + arr[j]
                if t == target:
                    sc, ec = 0, 0
                    s, e = arr[i], arr[j]
                    if s == e:
                        cnt += (j-i+1)*(j-i)//2
                        break
                    while i < len(arr) and arr[i] == s:
                        sc += 1
                        i += 1
                    while j >= 0 and arr[j] == e:
                        ec += 1
                        j -= 1
                    cnt += sc*ec
                elif t < target:
                    i += 1
                else:
                    j -= 1
            return cnt
            
        cnt = 0
        for i, e in enumerate(arr):
            cnt += count(i+1, target-e)
        return cnt


#{ 
 # Driver Code Starts
if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        arr = list(map(int, input().split()))
        target = int(input())
        ob = Solution()
        ans = ob.countTriplets(arr, target)
        print(ans)
        print("~")
# } Driver Code Ends

File: Difficulty: Medium/Count all triplets with given sum in sorted array/count-all-triplets-with-given-sum-in-sorted-array.java
//{ Driver Code Starts
// Initial Template for Java
import java.io.*;
import java.util.*;
import java.util.stream.Collectors;

public class Main {
    public static void main(String[] args) throws Exception {
        Scanner sc = new Scanner(System.in);
        int t = Integer.parseInt(sc.nextLine());
        while (t-- > 0) {
            String[] arr1Str = sc.nextLine().split(" ");
            int[] arr = Arrays.stream(arr1Str).mapToInt(Integer::parseInt).toArray();
            int target = Integer.parseInt(sc.nextLine());

            Solution ob = new Solution();
            int ans = ob.countTriplets(arr, target);
            System.out.println(ans);
            System.out.println("~");
        }
    }
}
// } Driver Code Ends


class Solution {
    public int countTriplets(int[] arr, int target) {
        // Code Here
       int n = arr.length;
        int count = 0;
       
        for (int i = 0; i < n; i++) {
            int x= i+1;
            int y=n-1;
            int z=target - arr[i];
            while(x<y){
                if(arr[x]+ arr[y] == z){
                    int right=y;
                    while(x<right && arr[right] == arr[y]){
                        count++;
                        right--;
                        
                    }
                    x++;
                }
                else if(arr[x]+arr[y]<z){
                    x++;
                }
                else{
                    y--;
                }
            }
        }
        return count;
                     
    }
}